patches/server/0977-Add-mob-spawner-ranges-changes-and-API.patch

```diff
@@ -0,0 +1,108 @@
++    public int spawnRangeVertical = GlobalConfiguration.get().misc.mobSpawnerRange.vertical; // Paper
+     private int tickDelay = 0; // Paper
+ 
+     public BaseSpawner() {}
+@@ -110,7 +113,7 @@ public abstract class BaseSpawner {
+                     ListTag nbttaglist = nbttagcompound.getList("Pos", 6);
+                     int j = nbttaglist.size();
+                     double d0 = j >= 1 ? nbttaglist.getDouble(0) : (double) pos.getX() + (randomsource.nextDouble() - randomsource.nextDouble()) * (double) this.spawnRange + 0.5D;
+-                    double d1 = j >= 2 ? nbttaglist.getDouble(1) : (double) (pos.getY() + randomsource.nextInt(3) - 1);
++                    double d1 = j >= 2 ? nbttaglist.getDouble(1) : (double) (pos.getY() + spawnRangeVertical == 0 ? 0 : (spawnRangeVertical < 0 ? randomsource.nextIntBetweenInclusive(spawnRangeVertical, -spawnRangeVertical) : randomsource.nextIntBetweenInclusive(-spawnRangeVertical, spawnRangeVertical))); // Paper
+                     double d2 = j >= 3 ? nbttaglist.getDouble(2) : (double) pos.getZ() + (randomsource.nextDouble() - randomsource.nextDouble()) * (double) this.spawnRange + 0.5D;
+ 
+                     if (world.noCollision(((EntityType) optional.get()).getAABB(d0, d1, d2))) {
+@@ -161,7 +164,7 @@ public abstract class BaseSpawner {
+                             return;
+                         }
+ 
+-                        int k = world.getEntitiesOfClass(entity.getClass(), (new AABB((double) pos.getX(), (double) pos.getY(), (double) pos.getZ(), (double) (pos.getX() + 1), (double) (pos.getY() + 1), (double) (pos.getZ() + 1))).inflate((double) this.spawnRange)).size();
++                        int k = world.getEntitiesOfClass(entity.getClass(), (new AABB((double) pos.getX(), (double) pos.getY(), (double) pos.getZ(), (double) (pos.getX() + 1), (double) (pos.getY() + 1), (double) (pos.getZ() + 1))).inflate((double) this.spawnRange, (double) spawnRangeVertical, (double) spawnRange)).size(); // Paper
+ 
+                         if (k >= this.maxNearbyEntities) {
+                             this.delay(world, pos);
+@@ -290,6 +293,12 @@ public abstract class BaseSpawner {
+         if (nbt.contains("SpawnRange", 99)) {
+             this.spawnRange = nbt.getShort("SpawnRange");
+         }
++    
++        // Paper start
++        if (nbt.contains("SpawnRangeVertical", 99)) {
++            this.spawnRangeVertical = nbt.getShort("SpawnRangeVertical");
++        }
++        // Paper end
+ 
+         this.displayEntity = null;
+     }
+@@ -313,6 +322,7 @@ public abstract class BaseSpawner {
+         nbt.putShort("MaxNearbyEntities", (short) this.maxNearbyEntities);
+         nbt.putShort("RequiredPlayerRange", (short) this.requiredPlayerRange);
+         nbt.putShort("SpawnRange", (short) this.spawnRange);
++        nbt.putShort("SpawnRangeVertical", (short) this.spawnRangeVertical); // Paper
```

Comment:
I think `Paper.SpawnRangeVertical` should be used instead

Reply:
What if someone changes it via a plugin?

Reply:
Sorry, I'm new to this whole business, guess you're right :D
I'll add it